docs/rfc/rfc-010-p2p-light-client.rst

```diff
@@ -0,0 +1,127 @@
+node/process for the p2p light client (e.g. all use of this will be as a
+client.) 
+
+The ability to run light clients using the RPC system will continue to be
+maintained.
+
+LibP2P
+~~~~~~
+
+While some aspects of the P2P light client implementation are orthogonal to
+LibP2P project, it's useful to think about the ways that these efforts may
+combine or interact.
+
+We expect to be able to leverage libp2p tools to provide some kind of service
+discovery for tendermint-based networks. This means that it will be possible
+for the p2p stack to easily identify specialized nodes, (e.g. light clients)
+thus obviating many of the design challenges with providing this feature in
+the context of the current stack.
+
+Similarly, libp2p makes it possible for project to be able back their non-Go
+light clients, without the major task of first implementing tendermint's p2p
+connection handling. We should identify if there exist users (e.g. the go IBC
+relayer, it's maintainers, and operators) who would be able to take advantage
+of p2p light client, before switching to libp2p. To our knowledge there are no
+external implementations of this p2p protocol, and it seems unlikely that a
+team would implement this directly ahead of its impending removal.
+
+User Cases
+~~~~~~~~~~
+
+This RFC makes a few assumptions about the use cases and users of light
+clients in tendermint.
+
+One of the most active use cases for light clients is in the implementation of
+the IBC relayer. Thus, we expect that providing P2P light clients might
+increase the reliability of relayers and reduce the cost of running a relayer,
+because relayer operators won't have to decide between rely on public RPC
+endpoints (unreliable) or running their own full nodes (expensive.) This
+also assumes that there are *no* other uses of the RPC in the relayer, and
+unless the relayers have the option of dropping all RPC use, it's unclear if a
+P2P light client will actually be able to successfully remove the dependency
+on the RPC system.
+
+Given that the primary relayer implementation, is hermes (rust,) it might be
+safe to deliver a version of tendermint that adds a light client rector in
+the full nodes, but that does not provide an implementation of a Go light
+client. This either means that the rust implementation would need support for
+the legacy P2P connection protocol or wait for the libp2p implementation.
+
+Client side light client (e.g. wallets, etc.) users may always want to use (a
+subset) of the RPC rather than connect to the P2P network for an ephemeral
+use.
+
+Discussion
+----------
+
+Implementation Questions
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Most of the complication in the is how to have a long lived light client node
+that *only* runs the light client reactor, as this raises a few questions:
+
+- would users specify a single P2P node to connect to when creating a light
+  client or would they also need/want to discover peers?
```

Comment:
The light client protocol relies on being connected to several nodes to be able to receive data from different sources. So if a light client is initially connected to a single node it will need some means of discovery to find other nodes

Reply:
sorry, I got the numbers wrong, but I think the question stands: should users specify the nodes that it talks to (as is the current model,) or should it be able to discover and select peers on its own (and if so using what selection mechanism?)

I guess the crux of my question is "currently operators of light clients must care about what nodes they connect to, should p2p light client operators also need/want to care about the nodes they connect to, or can we (modulo seed nodes, and perhaps persistent peers?) make it possible for these users to care less about the nodes they're connect to"

Reply:
Yeah that's a good question. I'd tend to think that 99% of our users don't care what specific nodes their light client is connected to and are happy to have such a discovery mechanism in place. There may be the 1% who for some reason want specific nodes but I don't think it's worth accommodating for. 

To answer your question on the selection mechanism, I guess the goal is to have the most broadest base of witnesses to be able to cross-check headers. Practically, it's not possible to connect with every node to ensure that the primary doesn't have 1/3+ voting power and isn't trying to fool you. So we should set some reasonable defaults (say 4 witnesses) which is the optimal amount to connect to. It would be great (but a nice to have) if we could occasionally rotate the peers we were connected with. I can't think of selection criteria we'd want to adopt that would bias us to connecting to one node over another.

Reply:
I've added an answer to this question.